### PR TITLE
Fixes #3512 - File descriptor is not released after zip file uploaded…

### LINKED ISSUE #3512: File descriptor is not released after zip file uploaded via jetty-client
Every time you upload a file, there will be one more file descriptor, which will not be released.

### PR DESCRIPTION
**Pull request by sbordet**
… via jetty-client.

#3512.

In case of multiple parts only the last iterator was closed.
Now, every part's iterator is closed.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

### CODE REVIEW

**Review by gregw — CHANGES_REQUESTED** [review]
I find the code rather confusing having to use generic Iterator and then have type checks for Closeable and Callback etc.    This is doubly confusing when the Iterator wraps another Iterator.

Would it be possible to replace the generic:
```java
  Iterator iterator = content.getIterator();
  ...
  if (iterator instanceof Closeable)
    ((Closeable)iterator).close();
``` 
with
```java
  MultipartIterator iterator = content.getMultipartIterator();
  ...
  iterator.close();
```
or something similar?

**Comment by sbordet** [comment]
@gregw I'm not sure I understand where exactly you suggest to change the code - there is no `getIterator()` method anywhere.

The iterator is to iterate over the chunks of content.

`Closeable` is for those iterator implementations that open resources and must close them.
The close is driven from external code in case of other failures not related to the content itself (e.g. the application aborted).

This is a part of Jetty `HttpClient` that was born simple (too simple) and had to evolve with a few tricks that I don't particularly like but had to exist to maintain API compatibility.

**Comment by gregw** [comment]
Sorry I meant `Iterator iterator = parts.iterator();` rather than `getIterator()`
So my question is can we used typed iterators rather than `instanceof` checks?   I would have thought so in some places at least as we are getting the iterator from Parts?

**Comment by sbordet** [comment]
@gregw still not following. `Parts` is not an existing class. I only see `parts.iterator()` on the server.
Line number?

**Comment by gregw** [comment]
OK then same thing with `ContentProvider`.  I've looked at a few implementation of that and they are pretty much all Closeable and Callbacks and have iterators that are also Closeable and Callbacks, so wouldn't it be better to have a `ContentProviderIterator` interface that was a Closeable Callback Iterator and avoid all the instanceof checks?

I'm not saying this must be done... No is an acceptable answer, but only if it is really the case that you will get arbitrary iterators that are not actually Closeable and/or Callbacks.  If all of them are (or almost all of them), then having a type rather than doing type logic would be a lot clearing.  It's just a bit confusing to follow the weakly typed code when there are so many Closeable/Callbacks that are not explicit.

**Comment by sbordet** [comment]
@gregw see #3756 and #4400.

To answer your question, of the 8 concrete `ContentProvider` implementations:

* `ByteBufferContentProvider`, `BytesContentProvider` and `StringContentProvider` don't need anything special.
* `PathContentProvider` and `InputStreamContentProvider` need an `Iterator` that implements `Closeable`.
* `DeferredContentProvider` needs an `Iterator` that implements `Callback` and `Synchronizable`.
* `OutputStreamContentProvider` is a wrapper for `DeferredContentProvider`.
* `MultiPartContentProvider` is a composite of all the others, so needs all the features.

So about half needs "special" features, but the other half does not.

To sum up, I think for 9.4.x we need to live with the current status of things and "sniff" whether `Iterator`s implement `Callback` or `Closeable` etc.
For 10.0.x we may review all this and make it better.

**Review by gregw — APPROVED** [review]
